ignore/src/walk.rs

```diff
@@ -1470,64 +1453,44 @@ impl Worker {
-                    // loop below will break and we'll go back to looking for
-                    // more work.
-                    self.waiting(true);
-                    self.quitting(true);
-                    while !self.is_quit_now() {
-                        let nwait = self.num_waiting();
-                        let nquit = self.num_quitting();
-                        // If the number of waiting workers dropped, then
-                        // abort our attempt to quit.
-                        if nwait < self.threads {
-                            break;
-                        }
-                        // If all workers are in this quit loop, then we
-                        // can stop.
-                        if nquit == self.threads {
-                            return None;
-                        }
-                        // Otherwise, spin.
-                    }
+                    // Repeat quit message to wake up sleeping threads, if
+                    // any. The domino effect will ensure that every thread
+                    // will quit.
+                    self.waiting();
+                    self.tx.send(Message::Quit).unwrap();
+                    return None;
                 }
-                Err(_) => {
-                    self.waiting(true);
-                    self.quitting(false);
-                    if self.num_waiting() == self.threads {
-                        for _ in 0..self.threads {
-                            self.tx.send(Message::Quit).unwrap();
-                        }
-                    } else {
-                        // You're right to consider this suspicious, but it's
-                        // a useful heuristic to permit producers to catch up
-                        // to consumers without burning the CPU. It is also
-                        // useful as a means to prevent burning the CPU if only
-                        // one worker is left doing actual work. It's not
-                        // perfect and it doesn't leave the CPU completely
-                        // idle, but it's not clear what else we can do. :-/
-                        thread::sleep(Duration::from_millis(1));
+                Err(TryRecvError::Empty) => {
+                    // If it was the last running thread, then no more work
+                    // can arrive, thus we can safely start quitting.
+                    // Otherwise, a thread may spawn new works to be done.
+                    if self.waiting() == 1 {
+                        // Every other thread is blocked at the next recv().
+
+                        // Send the initial quit message and quit.
+                        self.tx.send(Message::Quit).unwrap();
+                        return None;
                     }
+
+                    // Wait for next `Work` or `Quit` message.
+                    value = self.rx.recv().map_err(|_| TryRecvError::Disconnected);
```

Comment:
Why the `map_err` here instead of an `unwrap`?

Reply:
A few lines below I wrote an `unreachable!()` assertion that also covers `try_recv()` and this way `recv()` errors. So you will get a meaningful error message when panicking (although it can never occur).

Reply:
Although that `uncreachable()!` statement below is unnecessary, and we could use `unwrap` everywhere, I just wanted to explicitly distinguish `Err(TryRecvError::Empty)` from `Err(TryRecvError::Disconnected)` cases. In the future maybe `Disconnected` brach will be the quit.

Reply:
Right, yeah, I saw that. But then it becomes harder to distinguish between "channel somehow got disconnected from a bug" and "this particular blocking `recv` failed." That is, forwarding the error like this strips information from the inevitable backtrace as far as I can tell. I'd rather see it panic closer to where the error occurred.

Reply:
Yeah I like the idea of distinguishing the error cases. That part is good. :)

Reply:
> I'd rather see it panic closer to where the error occurred

I have no so much experience in Rust. It seems very reasonable.

On the other hand, what if you want to call `drop` on the channel intentionally and you want it to quit instead of panicking. Then it's better to forward the error instead of `unwrap`. And I think, knowing what `(try_)recv` panicked is secondary. The most important information is that it hase been disconnected.

But it's hard to reason about, which would be the better choice, because it cannot occur now.

Reply:
> And I think, knowing which (try_)recv panicked is secondary. The most important information is that it hase been disconnected.

If you use `recv().unwrap()` instead of `recv().map_err(...)`, then we get both. Where as with the `map_err`, you only know that the channel was disconnected.

> But it's hard to reason about, which would be the better choice, because it cannot occur now.

Right, _unless_ there is a bug. So basically, this is about choosing a failure mode that provides the most information and is thus easier to debug, if and when it does occur.